### PR TITLE
Initial mqtt port

### PR DESCRIPTION
**Pull request by jvalkeal**
- Porting mqtt module as source and sink apps.
- Fixes spring-cloud-stream-app-starters/app-starters-release#8
- Fixes spring-cloud-stream-app-starters/app-starters-release#13

### CODE REVIEW

File: README.adoc [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# MQTT Source and Sink
+
+To learn more about this application and the supported properties, please review the following link.
+
+## Rabbit Source
+include::spring-cloud-starter-stream-source-mqtt/README.adoc[]
+
+## Rabbit Sink
+include::spring-cloud-starter-stream-sink-mqtt/README.adoc[]
```
Comment (garyrussell) [inline README.adoc]: `s/Rabbit/MQTT/` (both titles)

File: mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+ * @author Janne Valkealahti
+ *
+ */
+@ConfigurationProperties("mqtt")
+public class MqttProperties {
+
+	/**
+	 * location of the mqtt broker(s) (comma-delimited list)
+	 */
+	private String url = "tcp://localhost:1883";
```
Comment (garyrussell) [inline mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java]: This should be `String[]` - Spring will parse the CDL - the factory setter takes `String[]` (varargs).

File: mqtt-app-starters-test-support/src/main/java/org/springframework/cloud/stream/app/test/mqtt/MqttRule.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		IMqttClient client = null;
+		try {
+			DefaultMqttPahoClientFactory connectionFactory = new DefaultMqttPahoClientFactory();
+			connectionFactory.setServerURIs("tcp://localhost:1883");
+			connectionFactory.setUserName("guest");
+			connectionFactory.setPassword("guest");
+			connectionFactory.setConnectionTimeout(1);
+			client = connectionFactory.getClientInstance("tcp://localhost:1883", "scdf-test-client");
+			client.connect(connectionFactory.getConnectionOptions());
+		} catch (Exception e) {
```
Comment (garyrussell) [inline mqtt-app-starters-test-support/src/main/java/org/springframework/cloud/stream/app/test/mqtt/MqttRule.java]: catch on new line.

File: mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+		DefaultMqttPahoClientFactory factory = new DefaultMqttPahoClientFactory();
+		factory.setServerURIs(mqttProperties.getUrl());
+		factory.setUserName(mqttProperties.getUsername());
+		factory.setPassword(mqttProperties.getPassword());
+		factory.setCleanSession(mqttProperties.isCleanSession());
+		factory.setConnectionTimeout(mqttProperties.getConnectionTimeout());
+		factory.setKeepAliveInterval(mqttProperties.getKeepAliveInterval());
+		if (ObjectUtils.nullSafeEquals(mqttProperties.getPersistence(), "file")) {
+			factory.setPersistence(new MqttDefaultFilePersistence(mqttProperties.getPersistenceDirectory()));
+		} else if (ObjectUtils.nullSafeEquals(mqttProperties.getPersistence(), "memory")) {
```
Comment (garyrussell) [inline mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttConfiguration.java]: else on new line

File: mqtt-app-starters-test-support/src/main/java/org/springframework/cloud/stream/app/test/mqtt/MqttRule.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import org.junit.runners.model.Statement;
+import org.springframework.integration.mqtt.core.DefaultMqttPahoClientFactory;
+
+/**
+ * Rule skipping tests if mqtt is not available from localhost with default settings.
+ *
+ * @author Janne Valkealahti
+ *
+ */
+public class MqttRule extends TestWatcher implements TestRule {
```
Comment (garyrussell) [inline mqtt-app-starters-test-support/src/main/java/org/springframework/cloud/stream/app/test/mqtt/MqttRule.java]: For consistency with other starters, this should really extend `AbstractExternalResourceTestSupport`.

**Review by garyrussell — CHANGES_REQUESTED** [review]

**Comment by sobychacko** [comment]
We might want to set the version of this project at 1.0.0.BUILD-SNAPSHOT as this is a new addition to app starters.

**Comment by jvalkeal** [comment]
@sobychacko This versioning is one of a weirdest things in apps. While it's true that `1.0.0` is probably more appropriate but as all apps currently have same version we haven't followed this approach that new app starts from `1.0.0`. Or is it a case that since initial `1.0.0` release we haven't added a singe new app?

**Comment by sobychacko** [comment]
@jvalkeal we had new apps in bacon, but because of some special circumstances, we needed to baseline everything to 1.2.0. But i think going forward, for new app starters, we should start from 1.0.0.

File: README.adoc [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# MQTT Source and Sink
+
+To learn more about this application and the supported properties, please review the following link.
```
Comment (artembilan) [inline README.adoc]: > Source and Sink

So, might be better to tell like: " these applications "

File: mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import org.springframework.integration.mqtt.core.MqttPahoClientFactory;
+import org.springframework.util.ObjectUtils;
+
+/**
+ * Generic mqtt configuration.
+ *
+ * @author Janne Valkealahti
+ *
+ */
+public class MqttConfiguration {
```
Comment (artembilan) [inline mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttConfiguration.java]: Why this isn't marked with the `@Configuration`?
Does it really work as is?

File: mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+ * @author Janne Valkealahti
+ *
+ */
+@ConfigurationProperties("mqtt")
+public class MqttProperties {
+
+	/**
+	 * location of the mqtt broker(s) (comma-delimited list)
+	 */
+	private String[] url = new String[] { "tcp://localhost:1883" };
```
Comment (artembilan) [inline mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java:36]: Some of the properties might be required - need particular validations to protect against misconfiguration

File: spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+	private MqttSinkProperties properties;
+
+	@Autowired
+	private MqttPahoClientFactory mqttClientFactory;
+
+	@Bean
+	@ServiceActivator(inputChannel = Sink.INPUT)
+	public MessageHandler mqttOutbound() {
+		MqttPahoMessageHandler messageHandler = new MqttPahoMessageHandler(properties.getClientId(), mqttClientFactory);
+		messageHandler.setAsync(true);
```
Comment (artembilan) [inline spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkConfiguration.java]: `this.properties.isAsync()` must be used.

File: spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkProperties.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+/**
+ * Properties for the Mqtt Sink.
+ *
+ * @author Janne Valkealahti
+ *
+ */
+@ConfigurationProperties("mqtt")
```
Comment (artembilan) [inline spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkProperties.java]: I wonder how does that work if we use the same prefix on different `@ConfigurationProperties` in the same app...
I mean that `@EnableConfigurationProperties({ MqttProperties.class, MqttSinkProperties.class })` and this `mqtt` prefix on the `MqttProperties` as well. 

Does it mean that Boot properly merge them into one set and then is able to distribute provided properties to proper classes?
Just confused a bit and would like to have some clarification.
Thanks

File: spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkProperties.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+public class MqttSinkProperties {
+
+	/**
+	 * identifies the client
+	 */
+	private String clientId = "scdf.mqtt.client.id.sink";
```
Comment (artembilan) [inline spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkProperties.java]: I still don't understand why do we mention SCDF in these apps?
That's good to demonstrate that our products works together and it is really in our interests to promote our products, but these SCSt apps are supposed to be working as a stand-along micorservices or as a part of some other platform.

File: spring-cloud-starter-stream-source-mqtt/pom.xml [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+	<dependencies>
+		<dependency>
+			<groupId>org.springframework.cloud.stream.app</groupId>
+			<artifactId>mqtt-app-starters-common</artifactId>
+			<version>1.2.1.BUILD-SNAPSHOT</version>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.integration</groupId>
+			<artifactId>spring-integration-java-dsl</artifactId>
```
Comment (artembilan) [inline spring-cloud-starter-stream-source-mqtt/pom.xml]: This dependency is out of use.

**Review by artembilan — CHANGES_REQUESTED** [review]
That's all from my side.
Nothing critical at a glance.
Just some concerns which require a bit more clarification.

File: spring-cloud-starter-stream-source-mqtt/pom.xml (hunk 12 above)
Comment (jvalkeal) [reply spring-cloud-starter-stream-source-mqtt/pom.xml]: Yeah, that was brought over from rabbit app, wonder why it's there?

File: spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkProperties.java (hunk 11 above)
Comment (jvalkeal) [reply spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkProperties.java]: When porting from xd, it uses xd in id so followed a pattern. Wonder what it should be?

File: spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkProperties.java (hunk 10 above)
Comment (jvalkeal) [reply spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkProperties.java]: Doesn't matter as boot will simply bind to both classes. We don't have shared `MqttProperties` anywhere and didn't want to duplicate same properties.

File: spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkConfiguration.java (hunk 9 above)
Comment (jvalkeal) [reply spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkConfiguration.java]: Good catch, I'll need to add that.

File: mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttConfiguration.java (hunk 7 above)
Comment (jvalkeal) [reply mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttConfiguration.java]: If it's imported `@Configuration' is not needed.

File: spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkProperties.java (hunk 10 above)
Comment (artembilan) [reply spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkProperties.java]: But you did. `Sink` and `Source` variants have many same properties...

File: mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java (hunk 8 above)
Comment (jvalkeal) [reply mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java:36]: We used validators in xd to fail fast but it looks like we dropped using those. Not sure what to do?

Comment (artembilan) [reply mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java:36]: Not sure what you mean about "dropped", but even mentioned `RabbitSourceProperties` has validation annotations

File: spring-cloud-starter-stream-source-mqtt/pom.xml (hunk 12 above)
Comment (artembilan) [reply spring-cloud-starter-stream-source-mqtt/pom.xml]: Yep! We don't need it there as well. Good catch!

File: mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttConfiguration.java (hunk 7 above)
Comment (jvalkeal) [reply mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttConfiguration.java]: I can add it, makes it less confusing.

File: spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkProperties.java (hunk 10 above)
Comment (jvalkeal) [reply spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkProperties.java]: Point was to separate connection properties as what are used in a shared config creating `mqttClientFactory`

File: mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+	 * 'memory' or 'file'
+	 */
+	private String persistence = "memory";
+
+	/**
+	 * Persistence directory
+	 */
+	private String persistenceDirectory = "/tmp/paho";
+
+	@Size(min = 1)
```
Comment (artembilan) [inline mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java:73]: Adding validation annotations requires in the latest Boot to add `@Validated` to the class level: https://github.com/spring-cloud-stream-app-starters/app-starters-release/issues/82

File: spring-cloud-starter-stream-sink-mqtt/src/test/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkTests.java [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+import static org.junit.Assert.assertNotNull;
+
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.cloud.stream.app.test.mqtt.MqttTestSupport;
+//import org.springframework.cloud.stream.binder.test.junit.rabbit.RabbitTestSupport;
```
Comment (artembilan) [inline spring-cloud-starter-stream-sink-mqtt/src/test/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkTests.java]: Redundant code

**Review by artembilan — CHANGES_REQUESTED** [review]
I don't have anything to nit-pick any more.

File: mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java (hunk 13 above)
Comment (jvalkeal) [reply mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java:73]: I can add it but weird as I got validation errors without it.

Comment (artembilan) [reply mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java:73]: @jvalkeal , I pointed you a GH issue originated this request.
Not my call to do that. That is a warn from Boot.

Comment (jvalkeal) [reply mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java:73]: ah, warning. ok I'll add it.

File: mqtt-app-dependencies/pom.xml [hunk 15]
```diff
@@ -0,0 +1,103 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+	<groupId>org.springframework.cloud.stream.app</groupId>
+	<artifactId>mqtt-app-dependencies</artifactId>
+	<version>1.2.1.BUILD-SNAPSHOT</version>
```
Comment (artembilan) [inline mqtt-app-dependencies/pom.xml:7]: There is a discussion on the Tensorflow PR that new app starter should have version as `1.0.0`.

@sobychacko , confirm, please

**Review by artembilan — APPROVED** [review]
I don't have objections on this PR any more.
Can be merged.

Tested via RabbitMQ MQTT plugin. 
I published message to the subscription queue and get answer in the artificially bound `foo` queue like:
```
Exchange 	output
Routing Key 	output
Redelivered 	○
Properties 	
priority:	0
delivery_mode:	2
headers:	
mqtt_retained:	false
mqtt_qos:	0
mqtt_topic:	mqtt-subscription-stream/client/id/sourceqos0
mqtt_duplicate:	false
contentType:	text/plain
content_type:	text/plain
Payload
60 bytes
Encoding: string
	

rty fghjkmhgjkm hgjk hgjkh gjkhgjk ghjk ghjk ghkj ghjk hgj k
```

Tested `sink` there as well via `amqp.topic` and `foo` queue bound via `stream.mqtt` routing key:
```
Exchange 	amq.topic
Routing Key 	stream.mqtt
Redelivered 	○
Properties 	
delivery_mode:	2
headers:	
x-mqtt-publish-qos:	1
x-mqtt-dup:	false
Payload
13 bytes
Encoding: string
	

fghfghfghfgdh
```

File: mqtt-app-dependencies/pom.xml (hunk 15 above)
Comment (jvalkeal) [reply mqtt-app-dependencies/pom.xml:7]: I still strongly disagree going with with any other version other than 1.2.x as all others are 1.2.x as well. I would just be confusing.

**Review by garyrussell — APPROVED** [review]

**Comment by garyrussell** [comment]
Merged as fbc87b2bf542265eb657a59386dcd0738d57e2bf with version 1.0.0, some minor checkstyle changes and switch to use AssertJ.

**Pull request closed without merging**